File: app/web/modules/recipes.js
(() => {
  'use strict';

  const DRAFT_KEY = 'ns_recipe_draft_v51';
  const bots = [
    ['auto_level', 'Auto Leveling'], ['auto_daily', 'Auto Daily'],
    ['auto_hunting', 'Auto Hunting'], ['eudemon', 'Eudemon'],
    ['circus', 'Circus'], ['yokai', 'Yokai'], ['yokai_minigame', 'Yokai Minigame'],
    ['shadow_war', 'Shadow War'], ['monster', 'Monster Hunter'],
    ['mission_s', 'Mission S'], ['clan_war', 'Clan War'], ['mission', 'Mission Farmer'],
  ];

  function creds() {
    try {
      const value = JSON.parse(sessionStorage.getItem('ns_quick_login') || 'null');
      return value?.user && value?.pass ? { username: value.user, password: value.pass } : null;
    } catch (_) { return null; }
  }
  function optionMarkup() {
    return bots.map(([value, label]) => `<option value="${value}">${label}</option>`).join('');
  }
  function refreshRow(row) {
    const kind = row.querySelector('.rs-kind').value;
    const bot = row.querySelector('.rs-bot').value;
    const mode = row.querySelector('.rs-mode');
    const value = row.querySelector('.rs-value');
    const mission = row.querySelector('.rs-mission');
    const isWait = kind === 'wait';

    row.querySelector('.rs-bot').classList.toggle('hidden', isWait);
    mode.classList.toggle('hidden', isWait);
    mission.classList.toggle('hidden', isWait || bot !== 'mission');

    const levelOption = [...mode.options].find(o => o.value === 'level_at_least');
    if (levelOption) levelOption.disabled = bot !== 'auto_level';
    if (bot !== 'auto_level' && mode.value === 'level_at_least') mode.value = 'cycles';

    value.min = 1;
    value.max = isWait ? 86400 : (mode.value === 'level_at_least' ? 200 : 5000);
    value.placeholder = isWait ? 'Seconds' : (mode.value === 'level_at_least' ? 'Target level' : 'Cycles / cap');
  }
  function stepRow(initial = {}) {
    const row = document.createElement('div');
    row.className = 'ns-recipe-step';
    row.innerHTML = `
      <select class="input-field rs-kind"><option value="bot">Bot</option><option value="wait">Wait</option></select>
      <select class="input-field rs-bot">${optionMarkup()}</select>
      <select class="input-field rs-mode">
        <option value="cycles">Cycles</option><option value="until_stop">Until completed</option><option value="level_at_least">Until level ≥</option>
      </select>
      <input class="input-field rs-value" type="number" min="1" value="5">
      <input class="input-field rs-mission hidden" type="text" maxlength="80" placeholder="Mission ID (e.g. msn_60)">
      <div class="rs-actions">
        <button class="btn btn-toggle rs-up" type="button" title="Move up">↑</button>
        <button class="btn btn-toggle rs-down" type="button" title="Move down">↓</button>
        <button class="btn btn-toggle rs-remove" type="button" title="Remove"><i class="fa-solid fa-xmark"></i></button>
      </div>`;

    row.querySelector('.rs-kind').value = initial.kind || 'bot';
    row.querySelector('.rs-bot').value = initial.bot_type || 'auto_level';
    row.querySelector('.rs-mode').value = initial.mode || 'cycles';
    row.querySelector('.rs-value').value = initial.kind === 'wait'
      ? (initial.seconds || 5)
      : (initial.mode === 'level_at_least' ? (initial.target_level || 80) : (initial.cycles || initial.max_cycles || 5));
    row.querySelector('.rs-mission').value = initial.params?.mission_id || '';

    for (const cls of ['.rs-kind', '.rs-bot', '.rs-mode']) {
      row.querySelector(cls).onchange = () => { refreshRow(row); saveDraft(); };
    }
    row.querySelector('.rs-value').oninput = saveDraft;
    row.querySelector('.rs-mission').oninput = saveDraft;
    row.querySelector('.rs-remove').onclick = () => { row.remove(); saveDraft(); };
    row.querySelector('.rs-up').onclick = () => {
      const previous = row.previousElementSibling;
      if (previous) row.parentElement.insertBefore(row, previous);
      saveDraft();
    };
    row.querySelector('.rs-down').onclick = () => {
      const next = row.nextElementSibling;
      if (next) row.parentElement.insertBefore(next, row);
      saveDraft();
    };
    refreshRow(row);
    return row;
  }
  function collect() {
    const steps = [...document.querySelectorAll('.ns-recipe-step')].map(row => {
      const kind = row.querySelector('.rs-kind').value;
      const raw = Math.max(1, parseInt(row.querySelector('.rs-value').value || '1', 10));
      if (kind === 'wait') return { kind: 'wait', seconds: Math.max(5, raw), label: 'Wait' };
      const bot = row.querySelector('.rs-bot').value;
      const mode = row.querySelector('.rs-mode').value;
      const params = {};
      if (bot === 'mission') params.mission_id = row.querySelector('.rs-mission').value.trim();
      return {
        kind: 'bot', bot_type: bot, mode,
        cycles: mode === 'cycles' ? raw : 1,
        target_level: mode === 'level_at_least' ? raw : null,
        max_cycles: mode === 'cycles' ? raw : (mode === 'level_at_least' ? 500 : Math.min(5000, raw || 500)),
        params,
      };
    });
    return { name: (document.getElementById('recipe-name')?.value || 'Automation Recipe').trim(), steps };
  }
  function saveDraft() {
    try { localStorage.setItem(DRAFT_KEY, JSON.stringify(collect())); } catch (_) {}
  }
  function loadDraft() {
    try { return JSON.parse(localStorage.getItem(DRAFT_KEY) || 'null'); } catch (_) { return null; }
  }
  async function dryRun() {
    const response = await fetch('/api/bot/recipe/dry-run', {
      method: 'POST', headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ recipe: collect() }),
    });
    const data = await response.json();
    if (!response.ok) throw new Error(data.detail || `HTTP ${response.status}`);
    return data.dry_run;
  }
  function renderDryRun(result) {
    const host = document.getElementById('recipe-dry-result');
    if (!host) return;
    const outline = (result.outline || []).map(step =>
      `<li><strong>${step.index}. ${step.label}</strong> · ${window.NinjaUI?.formatDuration?.(step.estimate_seconds) || step.estimate_seconds + 's'}${step.bounded ? '' : ' · safety cap'}</li>`
    ).join('');
    const warnings = (result.warnings || []).map(x => `<li>${x}</li>`).join('');
    host.innerHTML = `
      <strong>Estimate: ${window.NinjaUI?.formatDuration?.(result.estimate_seconds) || result.estimate_seconds + 's'}</strong>
      · cycle budget ${result.cycle_budget || 0}/5000
      <ol>${outline}</ol>
      ${warnings ? `<div class="ns-recipe-warnings"><strong>Notes</strong><ul>${warnings}</ul></div>` : ''}`;
  }
  function ensure() {
    if (document.getElementById('modal-recipes')) return;
    const scheduler = document.getElementById('btn-scheduler') || document.getElementById('btn-settings');
    if (!scheduler) return;

    const btn = document.createElement('button');
    btn.id = 'btn-recipes'; btn.className = 'btn btn-toggle'; btn.title = 'Automation Studio';
    btn.innerHTML = '<i class="fa-solid fa-wand-magic-sparkles"></i>';
    scheduler.insertAdjacentElement('afterend', btn);

    const modal = document.createElement('div');
    modal.id = 'modal-recipes'; modal.className = 'modal-overlay';
    modal.innerHTML = `
      <div class="modal-content glass-panel ns-recipe-modal">
        <div class="modal-header"><h3><i class="fa-solid fa-wand-magic-sparkles"></i> Automation Studio</h3><button class="btn-close" type="button">&times;</button></div>
        <div class="input-wrapper"><label>Recipe name</label><input id="recipe-name" class="input-field" value="Daily + Leveling"></div>
        <div id="recipe-steps" class="ns-recipe-steps"></div>
        <button id="recipe-add" class="btn btn-toggle" type="button">+ Add step</button>
        <div class="ns-recipe-schedule">
          <label>Start at (optional)<input id="recipe-at" type="datetime-local" class="input-field"></label>
          <label>Repeat<select id="recipe-repeat" class="input-field"><option value="0">Once</option><option value="21600">6 hours</option><option value="43200">12 hours</option><option value="86400">Daily</option></select></label>
        </div>
        <div id="recipe-dry-result" class="ns-recipe-note">Dry-run belum dijalankan.</div>
        <div class="ns-recipe-actions ns-sticky-actions"><button id="recipe-dry" class="btn btn-toggle" type="button">Dry Run</button><button id="recipe-start" class="btn btn-primary" type="button">Start Recipe</button></div>
      </div>`;
    document.body.appendChild(modal);

    const steps = modal.querySelector('#recipe-steps');
    const draft = loadDraft();
    if (draft?.steps?.length) {
      document.getElementById('recipe-name').value = draft.name || 'Automation Recipe';
      draft.steps.forEach(step => steps.append(stepRow(step)));
    } else {
      steps.append(
        stepRow({ kind: 'bot', bot_type: 'auto_daily', mode: 'until_stop', max_cycles: 100 }),
        stepRow({ kind: 'bot', bot_type: 'auto_level', mode: 'level_at_least', target_level: 80, max_cycles: 500 }),
      );
    }

    document.getElementById('recipe-name').addEventListener('input', saveDraft);
    btn.onclick = () => modal.classList.add('show');
    modal.querySelector('.btn-close').onclick = () => modal.classList.remove('show');
    modal.querySelector('#recipe-add').onclick = () => {
      if (steps.children.length >= 20) return window.NinjaUI?.toast('Maksimal 20 step.', 'warn');
      steps.append(stepRow()); saveDraft();
    };
    modal.querySelector('#recipe-dry').onclick = async () => {
      try { renderDryRun(await dryRun()); }
      catch (error) { window.NinjaUI?.toast(`Dry-run gagal: ${error.message}`, 'error'); }
    };
    modal.querySelector('#recipe-start').onclick = async () => {
      const s = window.NinjaSession?.get?.();
      if (!s?.sessionkey) return window.NinjaUI?.toast('Login terlebih dahulu.', 'warn');
      try {
        renderDryRun(await dryRun());
        saveDraft();
        const params = { recipe: collect() };
        const rawAt = document.getElementById('recipe-at').value;
        if (rawAt) {
          const date = new Date(rawAt);
          if (Number.isNaN(date.getTime()) || date.getTime() < Date.now() + 30000) throw new Error('Waktu start harus di masa depan.');
          params.schedule_at = Math.floor(date.getTime() / 1000);
        }
        const repeat = parseInt(document.getElementById('recipe-repeat').value || '0', 10) || 0;
        if (repeat) params.repeat_every_seconds = repeat;

        const response = await fetch('/api/bot/cloud/start', {
          method: 'POST', headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ sessionkey: s.sessionkey, char_id: s.char_id, bot_type: 'recipe', params, credentials: creds() }),
        });
        const data = await response.json();
        if (!response.ok) throw new Error(data.detail || `HTTP ${response.status}`);

        localStorage.setItem(`ns_cloud_control_${s.char_id}`, JSON.stringify({ char_id: s.char_id, token: data.job.control_token, bot_type: 'recipe' }));
        modal.classList.remove('show');
        window.dispatchEvent(new CustomEvent('ns:cloud-control'));
        window.dispatchEvent(new CustomEvent('ns:cloud-status', { detail: data.job }));
        window.NinjaRealtime?.connect?.();
      } catch (error) {
        window.NinjaUI?.toast(`Recipe gagal: ${error.message}`, 'error');
      }
    };
  }

  document.addEventListener('DOMContentLoaded', ensure);
  window.NinjaRecipes = { collect, dryRun };
})();

File: app/services/recipes.py
from __future__ import annotations

from typing import Any, Dict, List

ALLOWED_BOTS = {
    "auto_level", "auto_daily", "auto_hunting", "eudemon", "circus", "yokai",
    "yokai_minigame", "shadow_war", "monster", "mission_s", "clan_war", "mission",
}
ALLOWED_MODES = {"cycles", "until_stop", "level_at_least"}
MAX_STEPS = 20
MAX_CYCLES = 5000


def _safe_params(raw: Any) -> Dict[str, Any]:
    if not isinstance(raw, dict):
        return {}
    allowed = {"mission_id", "boss_type", "max_level"}
    return {str(k): v for k, v in raw.items() if k in allowed}


def validate_recipe(recipe: Any) -> Dict[str, Any]:
    if not isinstance(recipe, dict):
        raise ValueError("recipe must be an object")
    name = str(recipe.get("name") or "Automation Recipe").strip()[:80] or "Automation Recipe"
    raw_steps = recipe.get("steps")
    if not isinstance(raw_steps, list) or not raw_steps:
        raise ValueError("recipe.steps must contain at least one step")
    if len(raw_steps) > MAX_STEPS:
        raise ValueError(f"recipe supports at most {MAX_STEPS} steps")

    steps: List[Dict[str, Any]] = []
    for index, raw in enumerate(raw_steps, start=1):
        if not isinstance(raw, dict):
            raise ValueError(f"recipe step {index} must be an object")
        kind = str(raw.get("kind") or "bot").lower()
        if kind == "wait":
            seconds = int(raw.get("seconds") or 0)
            if seconds < 5 or seconds > 86400:
                raise ValueError(f"recipe wait step {index} must be 5..86400 seconds")
            steps.append({"kind": "wait", "seconds": seconds, "label": str(raw.get("label") or "Wait")[:80]})
            continue
        if kind != "bot":
            raise ValueError(f"unsupported recipe step kind: {kind}")

        bot_type = str(raw.get("bot_type") or "").strip().lower()
        if bot_type not in ALLOWED_BOTS:
            raise ValueError(f"unsupported recipe bot_type at step {index}: {bot_type}")
        mode = str(raw.get("mode") or "cycles").lower()
        if mode not in ALLOWED_MODES:
            raise ValueError(f"unsupported recipe mode at step {index}: {mode}")
        cycles = int(raw.get("cycles") or 1)
        if mode == "cycles" and not (1 <= cycles <= MAX_CYCLES):
            raise ValueError(f"recipe cycles at step {index} must be 1..{MAX_CYCLES}")
        target_level = raw.get("target_level")
        if mode == "level_at_least":
            target_level = int(target_level or 0)
            if target_level < 2 or target_level > 200:
                raise ValueError(f"target_level at step {index} must be 2..200")

        steps.append({
            "kind": "bot",
            "bot_type": bot_type,
            "mode": mode,
            "cycles": cycles if mode == "cycles" else 0,
            "target_level": target_level if mode == "level_at_least" else None,
            "max_cycles": max(1, min(MAX_CYCLES, int(raw.get("max_cycles") or (cycles if mode == "cycles" else 500)))),
            "params": _safe_params(raw.get("params")),
            "label": str(raw.get("label") or bot_type.replace("_", " ").title())[:80],
        })
    return {"name": name, "steps": steps, "version": 1}


def dry_run(recipe: Any, settings: Dict[str, Any]) -> Dict[str, Any]:
    normalized = validate_recipe(recipe)
    delays = {
        "auto_level": settings.get("leveling_delay_seconds", 5),
        "auto_daily": settings.get("daily_delay_seconds", 8),
        "auto_hunting": settings.get("hunting_delay_seconds", 8),
        "eudemon": settings.get("eudemon_delay_seconds", 10),
        "circus": settings.get("circus_delay_seconds", 10),
        "yokai": settings.get("yokai_delay_seconds", 10),
        "yokai_minigame": settings.get("yokai_minigame_delay_seconds", 8),
        "shadow_war": settings.get("shadow_war_between_battles_seconds", 30),
        "monster": settings.get("monster_delay_seconds", 8),
        "mission_s": settings.get("mission_s_delay_seconds", 8),
        "clan_war": settings.get("clan_war_battle_delay_seconds", 8),
        "mission": settings.get("mission_delay_seconds", 10),
    }
    seconds = 0
    unknown = False
    outline = []
    for idx, step in enumerate(normalized["steps"], start=1):
        if step["kind"] == "wait":
            seconds += step["seconds"]
            outline.append({"index": idx, "label": step["label"], "estimate_seconds": step["seconds"], "bounded": True})
            continue
        base = max(3, int(delays.get(step["bot_type"], 10) or 10))
        if step["mode"] == "cycles":
            estimate = base * step["cycles"]
            seconds += estimate
            bounded = True
        else:
            estimate = base * step["max_cycles"]
            seconds += estimate
            bounded = False
            unknown = True
        outline.append({"index": idx, "label": step["label"], "estimate_seconds": estimate, "bounded": bounded})
    warnings = [
        "Estimates use configured base delays; battle waits, resource waits, rate-limit backoff, exams and relogin can extend runtime."
    ]
    if unknown:
        warnings.append("At least one condition-driven step has no exact completion time; its max_cycles cap is used for the upper estimate.")
    return {"recipe": normalized, "estimate_seconds": seconds, "exact": not unknown, "outline": outline, "warnings": warnings}
